typescript/search-attributes.md

```diff
@@ -0,0 +1,86 @@
+## Search attributes V2
+
+### Why do we need this?
+
+In the current version of the SDK, search attributes are always arrays. In order for the server to support visibility
+on top of standard persistence, we need to redo the search attributes API to enforce single values for all search
+attributes except for keywords. The server will eventually drop support for multiple search attribute values for
+non-keywords.
+
+### Minimal proposal
+
+Make the E2E experience consistent, including query construction for listing workflows and search attribute
+registration.
+
+`workflows.ts`
+
+```ts
+import { searchAttributes } from '@temporalio/common';
```

Comment:
What does `searchAttributes` look like? Can I iterate all of them inside the workflow? Can I do things like unset all of the ones with a certain prefix? (i.e. a replace-all)

Reply:
This is just a set of functions and constant, it includes things like:

```
defineAttribute
WorkflowType // built-in attribute
```

Reply:
Can I iterate all of them inside the workflow? Can I do things like unset all of the ones with a certain prefix? (i.e. a replace-all)

Reply:
You'd have to define your own set of search attributes, it's up to you how to access those.
We can expose an API to list the set of search attributes, do you think it's valuable?
Since search attributes have to be registered, I don't know if there's a real need for it.

Reply:
TypeScript doesn't allow object destructuring on imports statements. That means that one would always have to write `searchAttributes.defineAttribute(...)`, instead of simply `defineAttribute(...)`.

I think `defineAttribute` should be outside of this object, and `searchAttributes` should contain only build-in search attributes...

Reply:
There's always:
```ts
const { defineAttribute, Keyword } = searchAttributes
```

Alternative (which I don't love):
```ts
import { defineAttribute, Keyword } from '@temporalio/common/lib/search-attributes'
```

Alternative 2 (possible) with some extra files:
```ts
import { defineAttribute, Keyword } from '@temporalio/common/search-attributes'
```

Alternative 3 (really don't love):
```
import { defineAttribute, Keyword } from '@temporalio/search-attributes'
```

Reply:
Maybe let's do (2)?

Reply:
Was thinking more about this:

```
import { defineAttribute, SearchAttributes } from '@temporalio/common';

const customerAttribute = defineAttribute('string', 'customer');

Q.filter(customerAttribute, 'johnny')
 .filter(SearchAttributes.workflowType, 'CustomerWorkflow');
```

Reply:
I like 2

Reply:
`SearchAttributes` is essentially an enum, from user's POV

Reply:
I like 2 as well.
Not sure why `SearchAttributes` are an enum..

Reply:
I think "built in search attributes" should be grouped together. Not formally an enum, but an object on which every field is a builtin search attribute.

Having an export named simply `WorkflowType` gives the impression that it is a type. Writing it with a lower case makes it only marginally better.

Reply:
Sure, maybe have a `builtins` constant with all of the builtin search attributes.